codersdk/toolsdk/toolsdk_test.go

```diff
@@ -376,11 +357,160 @@ func TestTools(t *testing.T) {
+			},
+		}
+
+		wrapped := toolsdk.WithRecover(fakeTool.Handler)
+		v, err := wrapped(context.Background(), toolsdk.Deps{}, []byte(`{}`))
+		require.NoError(t, err)
+		require.JSONEq(t, `{}`, string(v))
+	})
+
+	t.Run("Error", func(t *testing.T) {
+		t.Parallel()
+		fakeTool := toolsdk.GenericTool{
+			Tool: aisdk.Tool{
+				Name:        "fake_tool",
+				Description: "Returns an error for testing.",
+			},
+			Handler: func(ctx context.Context, tb toolsdk.Deps, args json.RawMessage) (json.RawMessage, error) {
+				return nil, assert.AnError
+			},
+		}
+		wrapped := toolsdk.WithRecover(fakeTool.Handler)
+		v, err := wrapped(context.Background(), toolsdk.Deps{}, []byte(`{}`))
+		require.Nil(t, v)
+		require.ErrorIs(t, err, assert.AnError)
+	})
+
+	t.Run("Panic", func(t *testing.T) {
+		t.Parallel()
+		panicTool := toolsdk.GenericTool{
+			Tool: aisdk.Tool{
+				Name:        "panic_tool",
+				Description: "Panics for testing.",
+			},
+			Handler: func(ctx context.Context, tb toolsdk.Deps, args json.RawMessage) (json.RawMessage, error) {
+				panic("you can't sweat this fever out")
+			},
+		}
+
+		wrapped := toolsdk.WithRecover(panicTool.Handler)
+		v, err := wrapped(context.Background(), toolsdk.Deps{}, []byte("disco"))
+		require.Empty(t, v)
+		require.ErrorContains(t, err, "you can't sweat this fever out")
+	})
+}
+
+type testContextKey struct{}
+
+func TestWithCleanContext(t *testing.T) {
+	t.Parallel()
+
+	t.Run("NoContextKeys", func(t *testing.T) {
+		t.Parallel()
+
+		// This test is to ensure that the context values are not set in the
+		// toolsdk package.
+		ctxTool := toolsdk.GenericTool{
+			Tool: aisdk.Tool{
+				Name:        "context_tool",
+				Description: "Returns the context value for testing.",
+			},
+			Handler: func(toolCtx context.Context, tb toolsdk.Deps, args json.RawMessage) (json.RawMessage, error) {
+				v := toolCtx.Value(testContextKey{})
+				assert.Nil(t, v, "expected the context value to be nil")
+				return nil, nil
+			},
+		}
+
+		wrapped := toolsdk.WithCleanContext(ctxTool.Handler)
+		ctx := context.WithValue(context.Background(), testContextKey{}, "test")
+		_, _ = wrapped(ctx, toolsdk.Deps{}, []byte(`{}`))
+	})
+
+	t.Run("PropagateCancel", func(t *testing.T) {
+		t.Parallel()
+
+		// This test is to ensure that the context is canceled properly.
+		ctxTool := toolsdk.GenericTool{
+			Tool: aisdk.Tool{
+				Name:        "context_tool",
+				Description: "Returns the context value for testing.",
+			},
+			Handler: func(toolCtx context.Context, tb toolsdk.Deps, args json.RawMessage) (json.RawMessage, error) {
+				// Wait for the context to be canceled
+				<-toolCtx.Done()
+				return nil, toolCtx.Err()
+			},
+		}
+		wrapped := toolsdk.WithCleanContext(ctxTool.Handler)
+		errCh := make(chan error, 1)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		t.Cleanup(cancel)
+		go func() {
+			_, err := wrapped(ctx, toolsdk.Deps{}, []byte(`{}`))
+			errCh <- err
+		}()
+
+		cancel()
+		select {
```

Comment:
c.f. `testutil.RequireReceive(t, ctx, errCh)`

Reply:
In this case `RequireReceive` will fail the test due to the canceled context.